GH-28: Added support for the super- and subtype hierarchies.

kittaakos opened the pull request:
~**Note:** This item depends on https://github.com/eclipse/lsp4j/issues/272 due to required API changes. Still, this PR is ready for review/preview.~ Done.

Closes: #28.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

![screencast 2018-10-11 15-18-32](https://user-images.githubusercontent.com/1405703/46811007-26757d80-cd72-11e8-981d-4e0606d5e600.gif)

<img width="482" alt="46809098-15c30880-cd6e-11e8-8146-fcdc90f6721b" src="https://user-images.githubusercontent.com/1405703/46811029-368d5d00-cd72-11e8-99fa-9b9069b18aee.png">

For reviewers:
 - ~I start working on a client implementation, once it is available, I will update this PR.~ A client implementation with Eclipse Theia is available [here](https://github.com/theia-ide/theia/pull/3148).
 - ~Why did I add `org.eclipse.wst.xml_ui.feature.feature.group` to the [TP](https://github.com/eclipse/eclipse.jdt.ls/pull/823/files#diff-876b543348991468f6ac3f66768a2973R40)~? The TP related change is not required anymore. See https://github.com/eclipse/eclipse.jdt.ls/pull/823#pullrequestreview-176602548.
   - LSP4J depends on Xtext/Xtend 2.15.
   - Xtext/Xtend 2.15 is not available from the `http://download.eclipse.org/releases/photon/` p2 site.
   - But `org.jboss.tools.maven.apt.feature.feature.group` depends on `org.eclipse.wst.xml.ui` that we used to pull from `http://download.eclipse.org/releases/photon/` under the hood.
   - I had to explicitly declare the `org.eclipse.wst.xml_ui.feature.feature.group` feature dependency to fulfill the missing `o.e.wst.xml.ui` plug-in.

Edit: Updated the section for the TP.

fbricon commented on org.eclipse.jdt.ls.target/org.eclipse.jdt.ls.tp.target:
```diff
@@ -33,14 +33,18 @@
         <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
-            <unit id="org.eclipse.xtend.sdk.feature.group" version="2.14.0.v20180523-0937"/>
-            <unit id="org.eclipse.xtext.sdk.feature.group" version="2.14.0.v20180523-0937"/>
+            <unit id="org.eclipse.wst.xml_ui.feature.feature.group" version="3.10.0.v201804210200"/>
```
should not be necessary anymore, since we now only add the m2e-apt core bundle to the TP https://github.com/eclipse/eclipse.jdt.ls/commit/74a09932d29772d4749e67ac00d3c5968e91e94c

kittaakos commented:
I have updated this PR based on the [feedback](https://github.com/Microsoft/vscode-languageserver-node/pull/346#issuecomment-440283344) we got for the LSP modifications.

I am currently experiencing build issues locally, it might be [related to Xtext](https://gitter.im/eclipse/xtext?at=5c476a46cb47ec300085d80d).

fbricon commented:
How can I test it?

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/TypeHierarchyHandler.java:
```diff
@@ -0,0 +1,305 @@
+
+	protected void resolveChildren(Collection<Pair<TypeHierarchyItem, IType>> pairs, int depth, IProgressMonitor monitor) throws JavaModelException {
+		if (depth > 0) {
+			for (Pair<TypeHierarchyItem, IType> pair : pairs) {
```
check cancelled monitor

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/TypeHierarchyHandler.java:
```diff
@@ -0,0 +1,305 @@
+
+	protected void resolveParents(Collection<Pair<TypeHierarchyItem, IType>> pairs, int depth, IProgressMonitor monitor) throws JavaModelException {
+		if (depth > 0) {
+			for (Pair<TypeHierarchyItem, IType> pair : pairs) {
```
check cancelled monitor

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentSymbolHandler.java:
```diff
@@ -228,7 +202,119 @@ private boolean isSyntheticElement(IJavaElement element) {
+	 * The location, encapsulating the range enclosing this symbol not including
+	 * leading/trailing whitespace but everything else like comments.
+	 */
+	public static Location getFullLocation(IJavaElement element) throws JavaModelException {
```
similar static methods related to location computation live in JDTUtils. I reckon it's getting bloated over time, so we might get a dedicated LocationUtils class to handle all that

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/TypeHierarchyHandler.java:
```diff
@@ -0,0 +1,305 @@
+			return null;
+		}
+		try {
+			Location fullLocation = DocumentSymbolHandler.getFullLocation(type);
```
should not reference DocumentSymbolHandler here. use utility class instead

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/TypeHierarchyHandler.java:
```diff
@@ -0,0 +1,305 @@
+
+	public TypeHierarchyHandler(PreferenceManager preferenceManager) {
+		this.preferenceManager = preferenceManager;
+		enabled = Suppliers.memoize(() -> this.preferenceManager.getClientPreferences().isTypeHierarchySupported());
```
not convinced this simple call need to be memoized

fbricon commented:
you might want to change copyright headers for new files to 2019.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java:
```diff
@@ -759,6 +762,14 @@ public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params) {
 
+	@Override
+	public CompletableFuture<TypeHierarchyItem> typeHierarchy(Either<TypeHierarchyParams, ResolveTypeHierarchyItemParams> params) {
+		logInfo(">> textDocument/typeHierarchy");
```
my understanding of the LSP proposal is we should have 2 separate requests, one to get the type hierarchy, another one to resolve type hierarchy items.  Similar to completion items. So we should have to implement 2 methods instead.

kittaakos commented:
> How can I test it?

It's not yet ready, sorry for the noise. I added the _[WIP]_ tag to the title. I had to push to see if the build failure is local or not.

I will update this PR once more with some hints on how to try it.

kittaakos commented:
> How can I test it?

You can build it from source, but you need Python 2.x, yarn, etc.. or you could try to open it in Gitpod.

1. [![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/theia-ide/theia/pull/3802)
2. Once the build is finished, you can start the modified version of Theia with the adjusted JDT LS.
Execute the `yarn --cwd ./examples/browser start --hostname 0.0.0.0 --port 3000` in the terminal.

![screen shot 2019-01-25 at 16 27 58](https://user-images.githubusercontent.com/1405703/51757115-dd4bfe00-20c2-11e9-8dd1-cb478c2e411e.jpg)

3. Click on `Open Browser` and [you have to wait a bit](https://github.com/gitpod-io/gitpod/issues/270).
![screen shot 2019-01-25 at 16 28 07](https://user-images.githubusercontent.com/1405703/51757134-e8069300-20c2-11e9-986d-5aa2b49fde46.jpg)

4. When you can see the welcome screen, you have to open a new terminal. It is either with <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>\`</kbd> or from the `Terminal` > `New Terminal` menu.

5. Clone the JDT LS source by executing `cd /workspace/ && git clone https://github.com/eclipse/eclipse.jdt.ls.git` in the new terminal. (You can get any other Java source.)

6. Open a new workspace.
![screen shot 2019-01-25 at 17 07 18](https://user-images.githubusercontent.com/1405703/51757470-bc37dd00-20c3-11e9-8fd0-e5ed0891c78d.jpg)
![screen shot 2019-01-25 at 17 07 26](https://user-images.githubusercontent.com/1405703/51757485-c35eeb00-20c3-11e9-9194-f144f0a119e3.jpg)
![screen shot 2019-01-25 at 17 07 35](https://user-images.githubusercontent.com/1405703/51757493-c954cc00-20c3-11e9-87f9-037ebfb8ff7e.jpg)

7. Once your workspace is up in Theia and the Java LS build has finished, open any Java files.
You can trigger the type hierarchy either with <kbd>Ctrl/Cmd</kbd>+<kbd>Alt</kbd>+<kbd>H</kbd> or from the editor's context menu.
![screen shot 2019-01-25 at 17 13 35](https://user-images.githubusercontent.com/1405703/51757773-93641780-20c4-11e9-9482-7ba4c7842aad.jpg)

Note, if the `Type Hierarchy` view is active, and has input, it will flip the type hierarchy direction.
![screencast 2019-01-25 17-15-17](https://user-images.githubusercontent.com/1405703/51758014-07062480-20c5-11e9-802f-91ef2395033a.gif)

If you have any question, or you stuck, let me know.

kittaakos commented:
An easier way; [here is Gitpod snapshot](https://gitpod.io/#snapshot/f8a01ef8-5c42-46f9-bc8f-0480b33f8a20)  containing all the steps I have described above.
After clicking the link, select `Open Browser` from the top, and you can open (<kbd>Ctrl/Cmd</kbd>+<kbd>P</kbd>) any Java files from the workspace to try out the feature.

kittaakos commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/TypeHierarchyHandler.java:
```diff
@@ -0,0 +1,316 @@
+	 */
+	private TypeHierarchyItem shallowCopy(TypeHierarchyItem other) {
+		TypeHierarchyItem copy = new TypeHierarchyItem();
+		other.setName(other.getName());
```
FIXME: must be `copy` instead of `other`.

kittaakos commented:
I have rebased this PR from the `master` and updated the LSP4J dependency from `0.6.0` to `0.7.0`. I also made the necessary API changes due to the LSP modifications. This PR is ready for the review.

Please note, although the proposed LSP extension has not been accepted by MS yet, this PR complies the [proposed changes](https://github.com/Microsoft/vscode-languageserver-node/pull/426).

yudar1024 commented:
wait this feature long time.

yudar1024 commented:
@fbricon can some one handle the confilict files? wait this feature long time?

fbricon commented:
@yudar1024 I want to wait for the spec to be finalized before merging this feature. See https://github.com/microsoft/vscode-languageserver-node/pull/426

yudar1024 commented:
@fbricon can you merge first to let developer can use it in early time like clangd, merge it at this time or after spec be finalize, we both need change the code if the spec make some change. but if we merge at this time , at least it can make vsc more efficient early

CsCherrYY commented:
Hi @kittaakos 

Are you still interested in this feature? I think it's especially useful and I'm interested in it. 

I found you closed [microsoft/vscode-languageserver-node#426](https://github.com/microsoft/vscode-languageserver-node/pull/426) which is a proposal of type hierarchy in LSP. Although it will not be available in LSP recently, VS Code allows extensions to contribute to the tree view and here is a [sample](https://code.visualstudio.com/api/extension-guides/tree-view). 

We could put the type hierarchy in the tree view, use a private command to exchange data instead of LSP currently. What do you think about it?

CsCherrYY commented:
@kittaakos @fbricon

I'm going to implement this feature and make it available first. In the future, we can move to LSP once it supports type hierarchy.

I notice the data models such as `TypeHierarchyItem`, `TypeHierarchyParams` and `ResolveTypeHierarchyItemParams` are available since [lsp4j#273](https://github.com/eclipse/lsp4j/pull/273). I'd like to start with these models and use a delegate command to communicate between client and server.

In VS Code side, [References Viewlet API](https://code.visualstudio.com/updates/v1_50#_references-viewlet-api) offers an API to contribute to the reference view. Type hierarchy data can be shown in the reference view with this, just like what call hierarchy works today.

Do you have any comment or suggestion on this?

fbricon commented:
Superseded by https://github.com/eclipse/eclipse.jdt.ls/pull/1656
